compiler/rustc_mir/src/transform/check_consts/validation.rs

```diff
@@ -440,16 +440,7 @@ impl Validator<'mir, 'tcx> {
                                 // arguments when determining importance.
                                 let kind = LocalKind::Arg;
 
-                                if constness == hir::Constness::Const {
-                                    self.check_op_spanned(ops::ty::TraitBound(kind), span);
-                                } else if !tcx.features().const_fn
-                                    || self.ccx.is_const_stable_const_fn()
-                                {
-                                    // HACK: We shouldn't need the conditional above, but trait
-                                    // bounds on containing impl blocks are wrongly being marked as
-                                    // "not-const".
-                                    self.check_op_spanned(ops::ty::TraitBound(kind), span);
-                                }
+                                self.check_op_spanned(ops::ty::TraitBound(kind), span);
                             }
                             // other kinds of bounds are either tautologies
                             // or cause errors in other passes
```

Comment:
Either the playground is overloaded or

```rust
#![feature(const_trait_impl)]
#![allow(incomplete_features)]

trait Foo<T> {
    fn foo();
}

impl const Foo<u32> for i32 {
    fn foo() {}
}

const fn foo<T>() where i32: Foo<T> {
    <i32 as Foo<T>>::foo()
}
```

hits this path and puts the compiler in an infinite loop.

Reply:
The playground doesn't have my patch though, so how would this indicate that removing the hack is wrong...?

Reply:
Also FWIW, [this builds fine for me](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=34c15ed741dbe268841c7954997252ca).